std/arith.asm

```diff
@@ -0,0 +1,178 @@
+    let CLK32_9 = |row| clock(9, row);
+    let CLK32_10 = |row| clock(10, row);
+    let CLK32_11 = |row| clock(11, row);
+    let CLK32_12 = |row| clock(12, row);
+    let CLK32_13 = |row| clock(13, row);
+    let CLK32_14 = |row| clock(14, row);
+    let CLK32_15 = |row| clock(15, row);
+    let CLK32_16 = |row| clock(16, row);
+    let CLK32_17 = |row| clock(17, row);
+    let CLK32_18 = |row| clock(18, row);
+    let CLK32_19 = |row| clock(19, row);
+    let CLK32_20 = |row| clock(20, row);
+    let CLK32_21 = |row| clock(21, row);
+    let CLK32_22 = |row| clock(22, row);
+    let CLK32_23 = |row| clock(23, row);
+    let CLK32_24 = |row| clock(24, row);
+    let CLK32_25 = |row| clock(25, row);
+    let CLK32_26 = |row| clock(26, row);
+    let CLK32_27 = |row| clock(27, row);
+    let CLK32_28 = |row| clock(28, row);
+    let CLK32_29 = |row| clock(29, row);
+    let CLK32_30 = |row| clock(30, row);
+    let CLK32_31 = |row| clock(31, row);
+    let CLK32 = [
+        CLK32_0, CLK32_1, CLK32_2, CLK32_3, CLK32_4, CLK32_5, CLK32_6, CLK32_7,
+        CLK32_8, CLK32_9, CLK32_10, CLK32_11, CLK32_12, CLK32_13, CLK32_14, CLK32_15,
+        CLK32_16, CLK32_17, CLK32_18, CLK32_19, CLK32_20, CLK32_21, CLK32_22, CLK32_23,
+        CLK32_24, CLK32_25, CLK32_26, CLK32_27, CLK32_28, CLK32_29, CLK32_30, CLK32_31
+    ];
+
+    /****
+    *
+    * LATCH POLS: x1,y1,x2,y2,y3
+    *
+    *****/
+
+    array::map(x1, |e| unchanged_until(e, CLK32[31]));
+    array::map(y1, |e| unchanged_until(e, CLK32[31]));
+    array::map(x2, |e| unchanged_until(e, CLK32[31]));
+    array::map(y2, |e| unchanged_until(e, CLK32[31]));
+    array::map(y3, |e| unchanged_until(e, CLK32[31]));
+
+    /****
+    *
+    * RANGE CHECK x1,y1,x2,y2,y3
+    *
+    *****/
+
+    sum(16, |i| x1[i] * CLK32[i]) + sum(16, |i| y1[i] * CLK32[16 + i]) in BYTE2;
+    sum(16, |i| x2[i] * CLK32[i]) in BYTE2;
+    sum(16, |i| y3[i] * CLK32[i]) + sum(16, |i| y2[i] * CLK32[16 + i]) in BYTE2;
+
+
+    /*******
+    *
+    * EQ0: A(x1) * B(y1) + C(x2) = D (y2) * 2 ** 256 + op (y3)
+    *        x1 * y1 + x2 - y2 * 2**256 - y3 = 0
+    *
+    *******/
+
+    /// returns a(0) * b(0) + ... + a(n - 1) * b(n - 1)
+    let dot_prod = |n, a, b| sum(n, |i| a(i) * b(i));
+    /// returns |n| a(0) * b(n) + ... + a(n) * b(0)
+    let product = |a, b| |n| dot_prod(n + 1, a, |i| b(n - i));
+    /// Converts array to function, extended by zeros.
+    let array_as_fun = [|arr| |i| if i < array::len(arr) {
+        if i >= 0 {
+            arr[i]
+        } else {
+            0
+        }
+    } else {
+        0
+    }][0];
+    let shift_right = |fn, amount| |i| fn(i - amount);
+
+    let x1f = array_as_fun(x1);
+    let y1f = array_as_fun(y1);
+    let x2f = array_as_fun(x2);
+    let y2f = array_as_fun(y2);
+    let y3f = array_as_fun(y3);
+
+    // Defined for arguments from 0 to 31 (inclusive)
+    let eq0 = (|| |nr|
+        product(x1f, y1f)(nr)
+        + x2f(nr)
+        - shift_right(y2f, 16)(nr)
```

Comment:
isn't that rather shift_left?

Reply:
I think not! `shift_right(y2f, 16)(i) = y2f(i - 16)`, so the element at index `i` is what used to be at index `i - 16`, so the array has been shifted right 16 times.

Perhaps what's confusing is that the limbs in `y2f` are little-endian, so right-shifting by 16 is equivalent to multiplying by `2**256`, which you would normally write as `<< 2**256`...

Reply:
Yeah exactly, the question is if you mean to shift the result or the function argument. For both you could argue that there is no intrinsic direction.

On the other hand, we use the term `product`, so at this point, we would take the view of the result.

Haha now I'm wondering if we would get exactly the same constraint if we replace `shift_right(y2f, 16)(nr)` by `product(y2f, two_pow_256)` and define a function `two_pow_256` accordingly. I guess it would work, since the optimizer removes multiplications by 0 and simplifies multiplications by one.